### ISSUE TITLE
Improve reStructuredText performance 

### ISSUE THREAD

**Issue by tychoish** [gohugoio/hugo#2651]
While we wait on a mature (or even moderately capable) native go reStructuredText support (e.g. #1436), I would like to open a new issue as a proposal for an intermediate solution. 

I'm preparing a site with ~1600 rst posts. Using the current implementation, it takes 2 and a half to three minutes to render the site. On a theory that we were loosing some time to python process startup, I prototyped a solution where: 
- a very basic flask app that takes a POST request with some and returns the rendered restructured task: 
  
  ``` python
  import docutils.core
  
  from flask import Flask
  from flask import request
  
  app = Flask(__name__)
  
  @app.route("/", methods=["POST", "GET"])
  def main():
      if request.method == "GET":
          return "rst2html service; use the POST method to convert"
      else:
          return docutils.core.publish_parts(request.data, writer_name="html")['body'].strip()
  
  if __name__ == '__main__':
      app.run()
  ```
- a modification to the existing `getRstContent()` function that does the appropriate HTTP request for converting the content.

This gets the rendering time down for my site down to ~30 seconds. I tried converting all the rst to markdown (which didn't go well,) and rendering was ~3 seconds, so this isn't ideal, and doesn't obviate the need for #1436, but puts rst support into the realm of usable. 

The challenge is how to make this solution production worthy. My initial inclination is to write the script above to a tempfile, and then start it in a goroutine in the background, but I'm not sure if hugo has a concept, of a "background job" that we could hook into, and by the time we need to render rst the first time, its too late. At the same time, rst clearly has a niche audience within hugo, and it seems disruptive to force RST to restructure the way that hugo processes text (though, I think that's not historically inconsistent, based on my memory of adding the existing rst support.) 

Would it make sense to write a "go rst library" with a functional interface, that uses this approach internally that hugo could consume? My largest concern is potentially leaking the flask app, but I assume that's possible to overcome. 

Is there another option that I'm missing?

**Comment by bep** [comment]
>The challenge is how to make this solution production worthy.

I would say that unless 1000s of people are shouting for speedier RST, your solution will never be accepted into Hugo. It adds complexity that I don't want to maintain/support.

You are right about the `os/exec` overhead -- if that could be reduced by batching the RST conversion somehow (it needs some refactoring, but Hugo could send all (or batches of) the pages in one Go to the external helper), that could be interesting. But that should also somehow include Aciidoc, which I guess is more popular than RST.

**Comment by bep** [comment]
That said, a "http page handler" fits nicely into a future plugin scenario, but the Python server would have to be provided by you.

**Comment by tychoish** [comment]
I've been thinking about writing a go package that does asciidoc/rst via a REST api running a mini "embedded" flask app, (but that would be an implementation detail hidden from hugo), but that would basically integrate into the existing implementation. 

Should be pretty easy to get something together, I'd implement it as something in my own account, and is this something that you'd be interested or willing to merge?

**Comment by bep** [comment]
>you'd be interested or willing to merge? 

We're not running any embedded "flask server" (whatever that is) inside Hugo, if that answered your question.

**Comment by tychoish** [comment]
Currently, hugo is willing to run third party python scripts (e.g. the current working asciidoc/rst implementation) via exec, and it cannot ensure these dependencies exist. 

I'm proposing a package that: 

- exposes a go functional interface to perform these text transformations.

- supports rst and asciidoc. 

- would be able to drop into the existing hugo implementation without architecture changes of any kind. 

I'm in favor of leaving the implementation details to that package and not discussing that as part of this proposal, though I think a more robust version of the script in the original post is pretty reasonable. 

It also makes some sense to me to create a package in pypi that depends on all of the Python dependencies, to simplify the user experience for getting access to these features.

**Comment by bep** [comment]
Let me think about it.

I agree on the tech side of your arguments, but it is still one more knob to care about. And that is a knob I don't need (Ascciidoctor, RST) personally. To take one example: I have spent a fair amount of energy to all the 3 template engines up to "Hugo spec", but I'm not even sure Ace or Amber is in use by any numbers. I have seen enough people contribute stuff and then leave the building, so I'm extremely hesitant to new knobs, esp. if they're not hitting the "core Hugo functionality".

**Comment by tychoish** [comment]
That's fair, though, I'm the original contributor of the restructured text support (and some other miscellany) and while I'm not hugely involved these days.

Also as envisioned, this wouldn't add any new knobs or potential configuration, but would just replace the existing functionality with a different backend, (admittedly one with more moving parts.)

**Comment by tychoish** [comment]
I'm willing to begin work on this soon, if the proposal makes sense to you (all?) If you have additional feedback, that would increase my chances of getting this changed merged, I would love to hear them.

**Comment by bep** [comment]
I'll delegate the decision making to @anthonyfok -- I believe he did the work on the Asciidoc side.

But it should:

* Replace any old solution, i.e. not several ways to generate RST / Asciidoc
* Any web server should be lazily started,  i.e. I, who doesn't use either of them, should not "feel it" outside the tests

**Comment by tychoish** [comment]
Agreed on replacing exiting old methods for generating rst/asciidoc. Method would be generalizable to any python text processing strategy, and lazily started.

What are the minimum supported go versions for hugo these days?

**Comment by bep** [comment]
>What are the minimum supported go versions for hugo these days?

1.6. We follow a two current-1 policy.

**Comment by tychoish** [comment]
I've created a new library package with a functional public interface that does what I've described above, that's https://github.com/tychoish/shimgo, the hugo impact is minimal comparatively and is in #2672 for consideration and further discussion.

**Comment by bep** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least six months.

The resources of the Hugo team are limited, and so we are asking for your help.

If this is a **bug** and you can still reproduce this error on the <code>master</code> branch, please reply with all of the information you have about it in order to keep the issue open.

If this is a **feature request**, and you feel that it is still relevant and valuable, please tell us why.

This issue will automatically be closed in the near future if no further activity occurs. Thank you for all your contributions.
